Anvil: Assethub Forking

jimjbrettj opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests.

Contributors guide: https://github.com/foundry-rs/foundry/blob/master/CONTRIBUTING.md

The contributors guide includes instructions for running rustfmt and building the
documentation.
-->

<!-- ** Please select "Allow edits from maintainers" in the PR Options ** -->

## Motivation

<!--
Explain the context and why you're making that change. What is the problem
you're trying to solve? In some cases there is not a problem and this can be
thought of as being the motivation for your change.
-->
Draft PR to show what I have currently for mocking inherent data and digests to support assethub forking
## Solution

<!--
Summarize the solution and provide any necessary context needed to understand
the code change.
-->

## PR Checklist

- [ ] Added Tests
- [ ] Added Documentation
- [ ] Breaking changes

iulianbarbu commented on crates/anvil-polkadot/src/substrate_node/service/mod.rs:
```diff
@@ -48,11 +85,312 @@ pub struct Service {
+    Ok(top_map)
+}
+
+fn build_forked_chainspec_from_raw_top(
```
Just to double check, since I think we might be able to simplify the setup here. Please note that I lack yet lots of context, so lmk if there are some existing decisions/assumptions my suggestions break.

1. I suspect we might use an outdated chain spec, when fetched over the network. Something like https://github.com/paritytech/polkadot-sdk/blob/master/cumulus/parachains/chain-specs/asset-hub-westend.json, which tbh, I doubt it points to latest version of assethub-westend. We can generate it based on the latest code in the runtime, with `staging-chain-spec-builder`: https://crates.io/crates/staging-chain-spec-builder. 

Build the runtime:
```
cargo build -p asset-hub-westend-runtime
```

And then:
```
chain-spec-builder create --raw-storage --relay-chain "westend-local" --runtime \
    target/release/wbuild/asset-hub-westend-runtime/asset_hub_westend_runtime.wasm named-preset development
```

2. At this moment we build the chain spec by fetching it over the network from an existing node. Can't we use the genesis chain spec generated like in 1 instead? Is it helpful for forking case to have a few blocks built and only that storage be used? Or maybe having the storage like at genesis means some keys are not initialized? Either way, that would be an interesting state to fork from, but yeh, maybe it is worth to work first on some state which has more keys initialized.

cc @alindima @jimjbrettj

skunert commented on crates/anvil-polkadot/src/substrate_node/service/mod.rs:
```diff
@@ -125,12 +463,35 @@ pub fn new(
+
+  //let para_id = Id::new(anvil_config.get_chain_id().try_into().unwrap());
+
+    let para_id = Id::new(420420421);
```
Maybe one problem you will face will be that this is not available because our internal substrate runtime does not implement it. In that case we need to introduce a fake_runtime_api here like in omni-node. This is done here for example [here](https://github.com/paritytech/polkadot-sdk/blob/skunert%2Fsync-free-finality-notifications/cumulus/polkadot-omni-node/lib/src/fake_runtime_api/mod.rs#L31-L31). The point of this is that the client does not know which runtime APIs are available in the WASM blob. So the fake runtime API tells is which ones are available. 

If the WASM blob in the end does not contain it, it will error.

iulianbarbu commented on crates/anvil-polkadot/src/substrate_node/service/mod.rs:
```diff
@@ -48,11 +69,228 @@ pub struct Service {
+        let next_block_number =
+            UniqueSaturatedInto::<u32>::unique_saturated_into(current_para_head.number) + 1;
+
+        let next_time = time_manager.current_call_timestamp().checked_mul(1000).unwrap();
```
We'll need to do: 
```suggestion
        let next_time = time_manager.next_timestamp().checked_mul(1000).unwrap();
```
The way we do it now will not advance time.

iulianbarbu commented on crates/anvil-polkadot/src/substrate_node/service/mod.rs:
```diff
@@ -48,11 +69,228 @@ pub struct Service {
+                    let http = HttpClientBuilder::default()
+                        .set_headers(headers)
+                        .build(http_url)
+                        .map_err(|e| eyre::eyre!("http client build error: {e}"))?;
```
I would replace this with subxt::RpcClient. Feels a bit more compact: https://docs.rs/subxt/latest/subxt/backend/rpc/struct.RpcClient.html.

iulianbarbu replied:
I actually thought we can also keep it around. The idea is simple - for fetching the entire state from an existing local chain. Would lazy loading backend allow it? This is only for the case where such thing would be useful of course, and since we have some code for it, it would be easy to just replace the lazy loading with a "sync" loading, when needed. But I am ok with also dropping the code.

iulianbarbu commented on crates/anvil-polkadot/src/substrate_node/service/mod.rs:
```diff
@@ -48,11 +69,228 @@ pub struct Service {
+                        .build(http_url)
+                        .map_err(|e| eyre::eyre!("http client build error: {e}"))?;
+                    let at_hex = resolve_fork_hash_http(&http, None).await?;
+                    let try_sync = fetch_sync_spec_http(&http, Some(at_hex.clone())).await;
```
My tests resulted in not being able to sync the spec. I think we might need to enable some experimental rpcs on the node, but I haven't figured it out yet. At the same time the way you did it, where we fetched all storage keys, works well, so we could rely only on that to simplify the code for now.

iulianbarbu commented on crates/anvil-polkadot/src/substrate_node/service/mod.rs:
```diff
@@ -48,11 +69,228 @@ pub struct Service {
     pub genesis_block_number: u64,
 }
 
+async fn resolve_fork_hash_http(
```
I would move all this new methods related to state fetching from an RPC node in a parachain we want to fork from in a different file. Better to start a directory called `fork` with a mod.rs, and we'll submodules, one being `fetch` maybe?, with all these new fns.

iulianbarbu commented on crates/anvil-polkadot/src/substrate_node/service/mod.rs:
```diff
@@ -48,11 +69,228 @@ pub struct Service {
+            current_para_block_head,
+            relay_offset: next_time as u32,
+            relay_blocks_per_para_block: 1,
+            para_blocks_per_relay_epoch: 10,
```
I experimented with these and I think it makes more sense to set `relay_blocks_per_para_block` and `para_blocks_per_relay_epoch` to 0, while setting the `relay_offset` to `LastRelayChainBlockNumber` (from `ParachainSystem` pallet) + 1. 

At the same time, to enable multiple blocks producing per same relay chain slot we should tweak the storage in terms of `RelaySlotInfo` from `aura-ext` pallet, and set the authored_in_relay part of the storage item to 0 with every inherents data providers creation. 

Example below (please note that I made some other changes so that I can have here a `BackendWithOverlay` instead of a plain backend):
<img width="818" height="58" alt="Image" src="https://github.com/user-attachments/assets/1b9e5ded-c48d-431e-936f-ae9e3cf2eb9c" />

All associated changes are visible here: https://github.com/paritytech/foundry-polkadot/commit/7e43e2d66f95b73de6d512284b82867b4f74d3a7. These changes allow us to build blocks <6s apart as well (besides the >=6s apart blocks, which worked in the past without the storage tweaking).

jimjbrettj replied:
updated!

jimjbrettj commented:
> One other important thing to note for all the comments left so far is that I've been able to consistently produce blocks 6 seconds apart with the storage fetching in the PR, and with the adjustments I suggested in the comments, from an asset-hub-westend network started with zombienet like so (copied from [ChainSafe#8](https://github.com/ChainSafe/foundry-polkadot/pull/8), and modified slightly):
> 
> ```
>     1. Install zombinets: https://github.com/paritytech/zombienet
>     2. build polkadot-parachain, polkadot, polkadot-execute-worker and polkadot-prepare-worker based on the polkadot-sdk commit found in the Cargo.lock of this branch, and ensure they are on the PATH.
>     3.  curl https://raw.githubusercontent.com/paritytech/polkadot-sdk/refs/heads/master/cumulus/zombienet/examples/asset_hub_westend_local_network.toml > asset-hub_westend_zombient.toml
>     4. modify the zombienet network spec from 3. in the `parachains` section, by specifying the `chain = "asset-hub-westend-dev"`.
>     5. Run zombienet zombienet spawn --provider native asset-hub_westend_zombient.toml
>     6. Take one collator RPC port (can be found associated to the polkadotjs url, emitted by zombienet, when 5. finishes)
>     7.  Run `cargo run -p anvil-polkadot -- --fork-url http://127.0.0.1:<port> --block-time 6`
> ```
> 
> I would say that shortly after addressing the comments we should look into the testing based on the mining strategies discussed.

Update: Step 7 should use was instead of HTTP, so `http://127.0.0.1:<port>` should be replaced with `ws://127.0.0.1:<port>`

iulianbarbu commented on crates/anvil-polkadot/Cargo.toml:
```diff
@@ -67,7 +67,20 @@ polkadot-sdk = { git = "https://github.com/paritytech/polkadot-sdk.git", branch
+hex = "0.4"
+indicatif.workspace = true
+
+
```
```suggestion
```

iulianbarbu commented:
@jimjbrettj , another issue you'll face, when sending transactions to the node to test the auto mining, is a panic, which stops the node and happens in our `sendTransaction` RPC handler, related to looking for the chain id in the storage, which doesn't exist in an asset-hub forked state, since the chain id there is constant and unmodifiable. 

We agreed some time ago that modifying the chain id will not work when forking, but sending transactions should still work by detecting we started the node from a fork, and use the chain id from the fork. Unfortunately there isn't a runtime API to get it, and the chain id is hardcoded in asset-hub-westend to 420_420_421. Not sure if the same chain id will be used for polkadot/kusama, or other assethub chains, but for now we shouldn't block with the testing in this, and just hardcode the chain id as well.

alindima commented:
> > > @jimjbrettj , another issue you'll face, when sending transactions to the node to test the auto mining, is a panic, which stops the node and happens in our `sendTransaction` RPC handler, related to looking for the chain id in the storage, which doesn't exist in an asset-hub forked state, since the chain id there is constant and unmodifiable.
> > > We agreed some time ago that modifying the chain id will not work when forking, but sending transactions should still work by detecting we started the node from a fork, and use the chain id from the fork. Unfortunately there isn't a runtime API to get it, and the chain id is hardcoded in asset-hub-westend to 420_420_421. Not sure if the same chain id will be used for polkadot/kusama, or other assethub chains, but for now we shouldn't block with the testing in this, and just hardcode the chain id as well.
> > 
> > 
> > hm it's part of the chain metadata, so can be queried through subxt if not found in storage
> 
> You mean the runtime metadata? It should be doable then to handle this inside the `chain_id` getter on the `ApiServer`.

Yes, I think it should be very easy

iulianbarbu commented:
Small update: through the edge cases I ran, I am quite confident that this logic is pretty much there against our mining strategies, delta the comments left so far. Here is a log excerpt where we can see a few things I tried. I tested individually and also mixed all mining modes: auto mining, interval based mining, and evm_mine RPC calls, while tweaking the time to past and future.

<details>
<summary>Excerpt</summary>

```log
2025-11-26T16:00:24.532386Z  INFO parachain-system: unincluded_segment_len: 1
2025-11-26T16:00:24.532430Z  INFO parachain-system: included_head: 0x32321a1a6ca98a71f3b8c315b67a52be51a42f1aba3b3455c8eff58b53de11d5
2025-11-26T16:00:24.534473Z  INFO parachain-system: ancestor that will be added to unincluded segment: <wasm:stripped>
2025-11-26T16:00:24.536003Z  INFO sc_basic_authorship::basic_authorship: 🎁 Prepared block for proposing at 15907 (5 ms) hash: 0x10a93bc7c66e3561b6d851ac5657d3dd77523866985fcf9b457dfb29486e617e; parent_hash: 0x3232…11d5; end: NoMoreTransactions; extrinsics_count: 2
2025-11-26T16:00:26.241870Z  INFO node::user: eth_chainId
2025-11-26T16:00:26.244188Z  INFO node::user: eth_getTransactionCount
2025-11-26T16:00:26.245545Z  INFO node::user: eth_gasPrice
2025-11-26T16:00:26.246660Z  INFO node::user: eth_estimateGas
2025-11-26T16:00:26.251102Z  INFO node::user: eth_sendRawTransaction
2025-11-26T16:00:26.252207Z  INFO node::user: eth_blockNumber
2025-11-26T16:00:26.252585Z  INFO sc_basic_authorship::basic_authorship: 🙌 Starting consensus session on top of parent 0x10a93bc7c66e3561b6d851ac5657d3dd77523866985fcf9b457dfb29486e617e (#15907)
2025-11-26T16:00:26.253044Z  INFO node::user: eth_getBlockByNumber
2025-11-26T16:00:26.253596Z  INFO node::user: eth_getTransactionReceipt
2025-11-26T16:00:26.254125Z  INFO parachain-system: unincluded_segment_len: 1
2025-11-26T16:00:26.254160Z  INFO parachain-system: included_head: 0x10a93bc7c66e3561b6d851ac5657d3dd77523866985fcf9b457dfb29486e617e
2025-11-26T16:00:26.257124Z  INFO node::user: eth_getTransactionReceipt
2025-11-26T16:00:26.257413Z  INFO parachain-system: ancestor that will be added to unincluded segment: <wasm:stripped>
2025-11-26T16:00:26.259214Z  INFO node::user: eth_getTransactionReceipt
2025-11-26T16:00:26.259549Z  INFO sc_basic_authorship::basic_authorship: 🎁 Prepared block for proposing at 15908 (6 ms) hash: 0xa908b6508aec2ec5dfc00b7ce390f03969d8783a080816cbb6b80701dace0db9; parent_hash: 0x10a9…617e; end: NoMoreTransactions; extrinsics_count: 3
2025-11-26T16:00:26.504798Z  INFO node::user: eth_blockNumber
2025-11-26T16:00:26.505854Z  INFO node::user: eth_getBlockByNumber
2025-11-26T16:00:26.508426Z  INFO node::user: eth_getTransactionReceipt
2025-11-26T16:00:30.480060Z  INFO node::user: eth_chainId
2025-11-26T16:00:30.482923Z  INFO node::user: eth_getTransactionCount
2025-11-26T16:00:30.484232Z  INFO node::user: eth_gasPrice
2025-11-26T16:00:30.485311Z  INFO node::user: eth_estimateGas
2025-11-26T16:00:30.490115Z  INFO node::user: eth_sendRawTransaction
2025-11-26T16:00:30.491041Z  INFO node::user: eth_blockNumber
2025-11-26T16:00:30.491468Z  INFO sc_basic_authorship::basic_authorship: 🙌 Starting consensus session on top of parent 0xa908b6508aec2ec5dfc00b7ce390f03969d8783a080816cbb6b80701dace0db9 (#15908)
2025-11-26T16:00:30.491728Z  INFO node::user: eth_getBlockByNumber
2025-11-26T16:00:30.492202Z  INFO node::user: eth_getTransactionReceipt
2025-11-26T16:00:30.492915Z  INFO parachain-system: unincluded_segment_len: 1
2025-11-26T16:00:30.492932Z  INFO parachain-system: included_head: 0xa908b6508aec2ec5dfc00b7ce390f03969d8783a080816cbb6b80701dace0db9
2025-11-26T16:00:30.493852Z  INFO node::user: eth_getTransactionReceipt
2025-11-26T16:00:30.495373Z  INFO parachain-system: ancestor that will be added to unincluded segment: <wasm:stripped>
2025-11-26T16:00:30.495554Z  INFO node::user: eth_getTransactionReceipt
2025-11-26T16:00:30.496727Z  INFO sc_basic_authorship::basic_authorship: 🎁 Prepared block for proposing at 15909 (5 ms) hash: 0x689e01773a2594324109856b89419fca11db657644d59ce3d4f35b4513f71644; parent_hash: 0xa908…0db9; end: NoMoreTransactions; extrinsics_count: 3
2025-11-26T16:00:30.743233Z  INFO node::user: eth_blockNumber
2025-11-26T16:00:30.744294Z  INFO node::user: eth_getBlockByNumber
2025-11-26T16:00:30.746530Z  INFO node::user: eth_getTransactionReceipt
2025-11-26T16:00:39.529955Z  INFO sc_basic_authorship::basic_authorship: 🙌 Starting consensus session on top of parent 0x689e01773a2594324109856b89419fca11db657644d59ce3d4f35b4513f71644 (#15909)
2025-11-26T16:00:39.532095Z  INFO parachain-system: unincluded_segment_len: 1
2025-11-26T16:00:39.532140Z  INFO parachain-system: included_head: 0x689e01773a2594324109856b89419fca11db657644d59ce3d4f35b4513f71644
2025-11-26T16:00:39.534835Z  INFO parachain-system: ancestor that will be added to unincluded segment: <wasm:stripped>
2025-11-26T16:00:39.536831Z  INFO sc_basic_authorship::basic_authorship: 🎁 Prepared block for proposing at 15910 (6 ms) hash: 0xcf19bcd151ae47ffaef0586be2f43b36e2f6096d863ca886b9d2a80c3818d0cb; parent_hash: 0x689e…1644; end: NoMoreTransactions; extrinsics_count: 2
2025-11-26T16:00:48.442006Z  INFO node::user: evm_setTime
2025-11-26T16:00:54.530452Z  INFO sc_basic_authorship::basic_authorship: 🙌 Starting consensus session on top of parent 0xcf19bcd151ae47ffaef0586be2f43b36e2f6096d863ca886b9d2a80c3818d0cb (#15910)
2025-11-26T16:00:54.532590Z  INFO parachain-system: unincluded_segment_len: 1
2025-11-26T16:00:54.532634Z  INFO parachain-system: included_head: 0xcf19bcd151ae47ffaef0586be2f43b36e2f6096d863ca886b9d2a80c3818d0cb
2025-11-26T16:00:54.534669Z  INFO parachain-system: ancestor that will be added to unincluded segment: <wasm:stripped>
2025-11-26T16:00:54.536142Z  INFO sc_basic_authorship::basic_authorship: 🎁 Prepared block for proposing at 15911 (5 ms) hash: 0x02b56f2897d63acc5df8ce298f43fdb1ba3863120b26090a03f71c70aae7b8cb; parent_hash: 0xcf19…d0cb; end: NoMoreTransactions; extrinsics_count: 2
2025-11-26T16:00:59.854868Z  INFO node::user: eth_chainId
2025-11-26T16:00:59.857098Z  INFO node::user: eth_getTransactionCount
2025-11-26T16:00:59.858309Z  INFO node::user: eth_gasPrice
2025-11-26T16:00:59.859629Z  INFO node::user: eth_estimateGas
2025-11-26T16:00:59.864147Z  INFO node::user: eth_sendRawTransaction
2025-11-26T16:00:59.865119Z  INFO node::user: eth_blockNumber
2025-11-26T16:00:59.865717Z  INFO sc_basic_authorship::basic_authorship: 🙌 Starting consensus session on top of parent 0x02b56f2897d63acc5df8ce298f43fdb1ba3863120b26090a03f71c70aae7b8cb (#15911)
2025-11-26T16:00:59.865885Z  INFO node::user: eth_getBlockByNumber
2025-11-26T16:00:59.866438Z  INFO node::user: eth_getTransactionReceipt
2025-11-26T16:00:59.867508Z  INFO parachain-system: unincluded_segment_len: 1
2025-11-26T16:00:59.867549Z  INFO parachain-system: included_head: 0x02b56f2897d63acc5df8ce298f43fdb1ba3863120b26090a03f71c70aae7b8cb
2025-11-26T16:00:59.867833Z  INFO node::user: eth_getTransactionReceipt
2025-11-26T16:00:59.869580Z  INFO node::user: eth_getTransactionReceipt
2025-11-26T16:00:59.871283Z  INFO parachain-system: ancestor that will be added to unincluded segment: <wasm:stripped>
2025-11-26T16:00:59.873569Z  INFO sc_basic_authorship::basic_authorship: 🎁 Prepared block for proposing at 15912 (7 ms) hash: 0x74e49c82929aaf1cf06607d4015c9c8c521b04068dfd54ddd4ce8f6406f1a1d1; parent_hash: 0x02b5…b8cb; end: NoMoreTransactions; extrinsics_count: 3
2025-11-26T16:01:00.117648Z  INFO node::user: eth_blockNumber
2025-11-26T16:01:00.118462Z  INFO node::user: eth_getBlockByNumber
2025-11-26T16:01:00.120164Z  INFO node::user: eth_getTransactionReceipt
2025-11-26T16:01:03.323858Z  INFO node::user: eth_chainId
2025-11-26T16:01:03.326228Z  INFO node::user: eth_getTransactionCount
2025-11-26T16:01:03.327507Z  INFO node::user: eth_gasPrice
2025-11-26T16:01:03.329254Z  INFO node::user: eth_estimateGas
2025-11-26T16:01:03.333838Z  INFO node::user: eth_sendRawTransaction
2025-11-26T16:01:03.335384Z  INFO sc_basic_authorship::basic_authorship: 🙌 Starting consensus session on top of parent 0x74e49c82929aaf1cf06607d4015c9c8c521b04068dfd54ddd4ce8f6406f1a1d1 (#15912)
2025-11-26T16:01:03.335611Z  INFO node::user: eth_blockNumber
2025-11-26T16:01:03.335732Z  INFO node::user: eth_getTransactionReceipt
2025-11-26T16:01:03.336400Z  INFO parachain-system: unincluded_segment_len: 1
2025-11-26T16:01:03.336422Z  INFO parachain-system: included_head: 0x74e49c82929aaf1cf06607d4015c9c8c521b04068dfd54ddd4ce8f6406f1a1d1
2025-11-26T16:01:03.337018Z  INFO node::user: eth_getBlockByNumber
2025-11-26T16:01:03.337710Z  INFO node::user: eth_getTransactionReceipt
2025-11-26T16:01:03.338342Z  INFO parachain-system: ancestor that will be added to unincluded segment: <wasm:stripped>
2025-11-26T16:01:03.339375Z  INFO sc_basic_authorship::basic_authorship: 🎁 Prepared block for proposing at 15913 (3 ms) hash: 0x6238e621473fdbb36bedfb6ec192ba39dc4aa48b4d955450bf3b55eaba98ebed; parent_hash: 0x74e4…a1d1; end: NoMoreTransactions; extrinsics_count: 3
2025-11-26T16:01:03.339735Z  INFO node::user: eth_getTransactionReceipt
2025-11-26T16:01:03.587946Z  INFO node::user: eth_blockNumber
2025-11-26T16:01:03.589144Z  INFO node::user: eth_getBlockByNumber
2025-11-26T16:01:03.590062Z  INFO node::user: eth_getTransactionReceipt
2025-11-26T16:01:09.119848Z  INFO node::user: evm_mine
2025-11-26T16:01:09.120499Z  INFO sc_basic_authorship::basic_authorship: 🙌 Starting consensus session on top of parent 0x6238e621473fdbb36bedfb6ec192ba39dc4aa48b4d955450bf3b55eaba98ebed (#15913)
2025-11-26T16:01:09.122191Z  INFO parachain-system: unincluded_segment_len: 1
2025-11-26T16:01:09.122223Z  INFO parachain-system: included_head: 0x6238e621473fdbb36bedfb6ec192ba39dc4aa48b4d955450bf3b55eaba98ebed
2025-11-26T16:01:09.124438Z  INFO parachain-system: ancestor that will be added to unincluded segment: <wasm:stripped>
2025-11-26T16:01:09.126085Z  INFO sc_basic_authorship::basic_authorship: 🎁 Prepared block for proposing at 15914 (5 ms) hash: 0xec1095e944e673b8cf36b3faa0f1d8f45e6330cfc8169c28475813fdcc25da97; parent_hash: 0x6238…ebed; end: NoMoreTransactions; extrinsics_count: 2
2025-11-26T16:01:09.129154Z  INFO node::user:     Block Number: 15914
2025-11-26T16:01:09.129171Z  INFO node::user:     Block Hash: 0xec1095e944e673b8cf36b3faa0f1d8f45e6330cfc8169c28475813fdcc25da97
2025-11-26T16:01:09.129188Z  INFO node::user:     Block Time: "Wed, 26 Nov 2025 15:13:40 +0000"

2025-11-26T16:01:09.529810Z  INFO sc_basic_authorship::basic_authorship: 🙌 Starting consensus session on top of parent 0xec1095e944e673b8cf36b3faa0f1d8f45e6330cfc8169c28475813fdcc25da97 (#15914)
2025-11-26T16:01:09.532165Z  INFO parachain-system: unincluded_segment_len: 1
2025-11-26T16:01:09.532213Z  INFO parachain-system: included_head: 0xec1095e944e673b8cf36b3faa0f1d8f45e6330cfc8169c28475813fdcc25da97
2025-11-26T16:01:09.534812Z  INFO parachain-system: ancestor that will be added to unincluded segment: <wasm:stripped>
2025-11-26T16:01:09.536618Z  INFO sc_basic_authorship::basic_authorship: 🎁 Prepared block for proposing at 15915 (6 ms) hash: 0xa383b4d1d4733d6450caa06009f077a29087ced855bf8959b7251b468eaca5a1; parent_hash: 0xec10…da97; end: NoMoreTransactions; extrinsics_count: 2
2025-11-26T16:01:24.529800Z  INFO sc_basic_authorship::basic_authorship: 🙌 Starting consensus session on top of parent 0xa383b4d1d4733d6450caa06009f077a29087ced855bf8959b7251b468eaca5a1 (#15915)
2025-11-26T16:01:24.531963Z  INFO parachain-system: unincluded_segment_len: 1
2025-11-26T16:01:24.532008Z  INFO parachain-system: included_head: 0xa383b4d1d4733d6450caa06009f077a29087ced855bf8959b7251b468eaca5a1
2025-11-26T16:01:24.534204Z  INFO parachain-system: ancestor that will be added to unincluded segment: <wasm:stripped>
2025-11-26T16:01:24.535583Z  INFO sc_basic_authorship::basic_authorship: 🎁 Prepared block for proposing at 15916 (5 ms) hash: 0xd3f5ac307659fb6f2870b9f38cff607a175d17e111118d324083f27dc4f1bea1; parent_hash: 0xa383…a5a1; end: NoMoreTransactions; extrinsics_count: 2
2025-11-26T15:06:52.065145Z  INFO eth-rpc::client: Node does not have getAutomine RPC. Defaulting to automine=false. error: Client(JsonRpc(ErrorObject { code: MethodNotFound, message: "Method not found", data: None }))
2025-11-26T15:06:52.065285Z  INFO eth-rpc::client: 🔌 Subscribing to new blocks (BestBlocks)
2025-11-26T15:06:52.065439Z  INFO eth-rpc::client: 🔌 Subscribing to new blocks (FinalizedBlocks)
2025-11-26T15:07:07.031013Z  INFO sc_basic_authorship::basic_authorship: 🙌 Starting consensus session on top of parent 0xe95f65508e016d69c8f2fb68dbbeec2cbf472a84f526c4cd30368da23023b77e (#15513)
2025-11-26T15:07:07.033645Z  INFO parachain-system: unincluded_segment_len: 3
2025-11-26T15:07:07.033681Z  INFO parachain-system: included_head: 0xe95f65508e016d69c8f2fb68dbbeec2cbf472a84f526c4cd30368da23023b77e
2025-11-26T15:07:07.036649Z  INFO parachain-system: ancestor that will be added to unincluded segment: <wasm:stripped>
2025-11-26T15:07:07.038224Z  INFO sc_basic_authorship::basic_authorship: 🎁 Prepared block for proposing at 15514 (6 ms) hash: 0x70f7a67cee782d08e3637efb17c2ee09446e0c77c40725f721848bfa4f295751; parent_hash: 0xe95f…b77e; end: NoMoreTransactions; extrinsics_count: 2
2025-11-26T15:07:22.031007Z  INFO sc_basic_authorship::basic_authorship: 🙌 Starting consensus session on top of parent 0x70f7a67cee782d08e3637efb17c2ee09446e0c77c40725f721848bfa4f295751 (#15514)
2025-11-26T15:07:22.033153Z  INFO parachain-system: unincluded_segment_len: 1
2025-11-26T15:07:22.033193Z  INFO parachain-system: included_head: 0x70f7a67cee782d08e3637efb17c2ee09446e0c77c40725f721848bfa4f295751
2025-11-26T15:07:22.036203Z  INFO parachain-system: ancestor that will be added to unincluded segment: <wasm:stripped>
2025-11-26T15:07:22.038410Z  INFO sc_basic_authorship::basic_authorship: 🎁 Prepared block for proposing at 15515 (7 ms) hash: 0x8ef97bfb2f4557f59dbfc573807cfe1dd371c9680557ce8fb447e381d7975c0f; parent_hash: 0x70f7…5751; end: NoMoreTransactions; extrinsics_count: 2
2025-11-26T15:07:37.031179Z  INFO sc_basic_authorship::basic_authorship: 🙌 Starting consensus session on top of parent 0x8ef97bfb2f4557f59dbfc573807cfe1dd371c9680557ce8fb447e381d7975c0f (#15515)
2025-11-26T15:07:37.033729Z  INFO parachain-system: unincluded_segment_len: 1
2025-11-26T15:07:37.033783Z  INFO parachain-system: included_head: 0x8ef97bfb2f4557f59dbfc573807cfe1dd371c9680557ce8fb447e381d7975c0f
2025-11-26T15:07:37.036636Z  INFO parachain-system: ancestor that will be added to unincluded segment: <wasm:stripped>
2025-11-26T15:07:37.038251Z  INFO sc_basic_authorship::basic_authorship: 🎁 Prepared block for proposing at 15516 (6 ms) hash: 0xe6be982e4a54c5707111b8265fb887cee1714157d29b0c14783c71cd0607cf30; parent_hash: 0x8ef9…5c0f; end: NoMoreTransactions; extrinsics_count: 2
2025-11-26T15:07:52.031157Z  INFO sc_basic_authorship::basic_authorship: 🙌 Starting consensus session on top of parent 0xe6be982e4a54c5707111b8265fb887cee1714157d29b0c14783c71cd0607cf30 (#15516)
2025-11-26T15:07:52.033247Z  INFO parachain-system: unincluded_segment_len: 1
2025-11-26T15:07:52.033291Z  INFO parachain-system: included_head: 0xe6be982e4a54c5707111b8265fb887cee1714157d29b0c14783c71cd0607cf30
2025-11-26T15:07:52.036047Z  INFO parachain-system: ancestor that will be added to unincluded segment: <wasm:stripped>
2025-11-26T15:07:52.037534Z  INFO sc_basic_authorship::basic_authorship: 🎁 Prepared block for proposing at 15517 (6 ms) hash: 0x13cdb34aa26d32c9b5ec0f1cbd64071b986b4de4a01174b66d78d143c01abf2d; parent_hash: 0xe6be…cf30; end: NoMoreTransactions; extrinsics_count: 2
2025-11-26T15:08:07.031298Z  INFO sc_basic_authorship::basic_authorship: 🙌 Starting consensus session on top of parent 0x13cdb34aa26d32c9b5ec0f1cbd64071b986b4de4a01174b66d78d143c01abf2d (#15517)
2025-11-26T15:08:07.033660Z  INFO parachain-system: unincluded_segment_len: 1
2025-11-26T15:08:07.033696Z  INFO parachain-system: included_head: 0x13cdb34aa26d32c9b5ec0f1cbd64071b986b4de4a01174b66d78d143c01abf2d
2025-11-26T15:08:07.036603Z  INFO parachain-system: ancestor that will be a2025-11-26T16:01:39.530149Z  INFO sc_basic_authorship::basic_authorship: 🙌 Starting consensus session on top of parent 0xd3f5ac307659fb6f2870b9f38cff607a175d17e111118d324083f27dc4f1bea1 (#15916)
2025-11-26T16:01:39.532229Z  INFO parachain-system: unincluded_segment_len: 1
2025-11-26T16:01:39.532268Z  INFO parachain-system: included_head: 0xd3f5ac307659fb6f2870b9f38cff607a175d17e111118d324083f27dc4f1bea1
2025-11-26T16:01:39.534292Z  INFO parachain-system: ancestor that will be added to unincluded segment: <wasm:stripped>
2025-11-26T16:01:39.535845Z  INFO sc_basic_authorship::basic_authorship: 🎁 Prepared block for proposing at 15917 (5 ms) hash: 0xd89001646cef5dbc1d153afd8d8a45f4ff20e7add04b177616cbf636d2b64839; parent_hash: 0xd3f5…bea1; end: NoMoreTransactions; extrinsics_count: 2
2025-11-26T16:01:54.529557Z  INFO sc_basic_authorship::basic_authorship: 🙌 Starting consensus session on top of parent 0xd89001646cef5dbc1d153afd8d8a45f4ff20e7add04b177616cbf636d2b64839 (#15917)
2025-11-26T16:01:54.531662Z  INFO parachain-system: unincluded_segment_len: 1
2025-11-26T16:01:54.531712Z  INFO parachain-system: included_head: 0xd89001646cef5dbc1d153afd8d8a45f4ff20e7add04b177616cbf636d2b64839
2025-11-26T16:01:54.533889Z  INFO parachain-system: ancestor that will be added to unincluded segment: <wasm:stripped>
2025-11-26T16:01:54.535460Z  INFO sc_basic_authorship::basic_authorship: 🎁 Prepared block for proposing at 15918 (5 ms) hash: 0x045ffd9576f0f108dc7c87351b7f1136edfd9c1ecc806c7f5e483cc6366d1e52; parent_hash: 0xd890…4839; end: NoMoreTransactions; extrinsics_count: 2
2025-11-26T16:02:09.530157Z  INFO sc_basic_authorship::basic_authorship: 🙌 Starting consensus session on top of parent 0x045ffd9576f0f108dc7c87351b7f1136edfd9c1ecc806c7f5e483cc6366d1e52 (#15918)
2025-11-26T16:02:09.532484Z  INFO parachain-system: unincluded_segment_len: 1
2025-11-26T16:02:09.532528Z  INFO parachain-system: included_head: 0x045ffd9576f0f108dc7c87351b7f1136edfd9c1ecc806c7f5e483cc6366d1e52
2025-11-26T16:02:09.534546Z  INFO parachain-system: ancestor that will be added to unincluded segment: <wasm:stripped>
2025-11-26T16:02:09.536058Z  INFO sc_basic_authorship::basic_authorship: 🎁 Prepared block for proposing at 15919 (5 ms) hash: 0x9d409066ecff816cccd975adf74aeee0657cf0bb8a856f9efc7a31ef7b9cc73a; parent_hash: 0x045f…1e52; end: NoMoreTransactions; extrinsics_count: 2
2025-11-26T16:02:24.530113Z  INFO sc_basic_authorship::basic_authorship: 🙌 Starting consensus session on top of parent 0x9d409066ecff816cccd975adf74aeee0657cf0bb8a856f9efc7a31ef7b9cc73a (#15919)
2025-11-26T16:02:24.532259Z  INFO parachain-system: unincluded_segment_len: 1
2025-11-26T16:02:24.532303Z  INFO parachain-system: included_head: 0x9d409066ecff816cccd975adf74aeee0657cf0bb8a856f9efc7a31ef7b9cc73a
2025-11-26T16:02:24.534360Z  INFO parachain-system: ancestor that will be added to unincluded segment: <wasm:stripped>
2025-11-26T16:02:24.535831Z  INFO sc_basic_authorship::basic_authorship: 🎁 Prepared block for proposing at 15920 (5 ms) hash: 0x0c599dd455fd4ebe50f4491b6bd466d10a1acebb4142d65360eebf93ef83f5c2; parent_hash: 0x9d40…c73a; end: NoMoreTransactions; extrinsics_count: 2
2025-11-26T16:02:35.487857Z  INFO node::user: evm_setTime
2025-11-26T16:02:39.529367Z  INFO sc_basic_authorship::basic_authorship: 🙌 Starting consensus session on top of parent 0x0c599dd455fd4ebe50f4491b6bd466d10a1acebb4142d65360eebf93ef83f5c2 (#15920)
2025-11-26T16:02:39.531774Z  INFO parachain-system: unincluded_segment_len: 1
2025-11-26T16:02:39.531827Z  INFO parachain-system: included_head: 0x0c599dd455fd4ebe50f4491b6bd466d10a1acebb4142d65360eebf93ef83f5c2
2025-11-26T16:02:39.534069Z  INFO parachain-system: ancestor that will be added to unincluded segment: <wasm:stripped>
2025-11-26T16:02:39.535588Z  INFO sc_basic_authorship::basic_authorship: 🎁 Prepared block for proposing at 15921 (5 ms) hash: 0x14ab4040932aa9946df79955cdf5e8f66a7ae0499c8c4742b4c3f5a19cc1e121; parent_hash: 0x0c59…f5c2; end: NoMoreTransactions; extrinsics_count: 2
2025-11-26T16:02:40.937600Z  INFO node::user: eth_chainId
2025-11-26T16:02:40.939840Z  INFO node::user: eth_getTransactionCount
2025-11-26T16:02:40.941169Z  INFO node::user: eth_gasPrice
2025-11-26T16:02:40.942184Z  INFO node::user: eth_estimateGas
2025-11-26T16:02:40.946437Z  INFO node::user: eth_sendRawTransaction
2025-11-26T16:02:40.947550Z  INFO node::user: eth_blockNumber
2025-11-26T16:02:40.948118Z  INFO sc_basic_authorship::basic_authorship: 🙌 Starting consensus session on top of parent 0x14ab4040932aa9946df79955cdf5e8f66a7ae0499c8c4742b4c3f5a19cc1e121 (#15921)
2025-11-26T16:02:40.948168Z  INFO node::user: eth_getBlockByNumber
2025-11-26T16:02:40.948699Z  INFO node::user: eth_getTransactionReceipt
2025-11-26T16:02:40.950008Z  INFO parachain-system: unincluded_segment_len: 1
2025-11-26T16:02:40.950047Z  INFO node::user: eth_getTransactionReceipt
2025-11-26T16:02:40.950060Z  INFO parachain-system: included_head: 0x14ab4040932aa9946df79955cdf5e8f66a7ae0499c8c4742b4c3f5a19cc1e121
2025-11-26T16:02:40.951822Z  INFO node::user: eth_getTransactionReceipt
2025-11-26T16:02:40.953854Z  INFO parachain-system: ancestor that will be added to unincluded segment: <wasm:stripped>
2025-11-26T16:02:40.956043Z  INFO sc_basic_authorship::basic_authorship: 🎁 Prepared block for proposing at 15922 (7 ms) hash: 0x29b6eb465b7dcbabf9b8fce05d84e3b2d834d834842ab96332f2576ab7b69992; parent_hash: 0x14ab…e121; end: NoMoreTransactions; extrinsics_count: 3
2025-11-26T16:02:41.199868Z  INFO node::user: eth_blockNumber
2025-11-26T16:02:41.201471Z  INFO node::user: eth_getBlockByNumber
2025-11-26T16:02:41.202281Z  INFO node::user: eth_getTransactionReceipt
2025-11-26T16:02:45.680210Z  INFO node::user: evm_mine
2025-11-26T16:02:45.680628Z  INFO sc_basic_authorship::basic_authorship: 🙌 Starting consensus session on top of parent 0x29b6eb465b7dcbabf9b8fce05d84e3b2d834d834842ab96332f2576ab7b69992 (#15922)
2025-11-26T16:02:45.682270Z  INFO parachain-system: unincluded_segment_len: 1
2025-11-26T16:02:45.682300Z  INFO parachain-system: included_head: 0x29b6eb465b7dcbabf9b8fce05d84e3b2d834d834842ab96332f2576ab7b69992
2025-11-26T16:02:45.683671Z  INFO parachain-system: ancestor that will be added to unincluded segment: <wasm:stripped>
2025-11-26T16:02:45.684667Z  INFO sc_basic_authorship::basic_authorship: 🎁 Prepared block for proposing at 15923 (3 ms) hash: 0x0dfaf1d789b56b179e4ff0ad0d4e0585806c1b31236606bc939c5deb61c67e54; parent_hash: 0x29b6…9992; end: NoMoreTransactions; extrinsics_count: 2
2025-11-26T16:02:45.688188Z  INFO node::user:     Block Number: 15923
2025-11-26T16:02:45.688198Z  INFO node::user:     Block Hash: 0x0dfaf1d789b56b179e4ff0ad0d4e0585806c1b31236606bc939c5deb61c67e54
2025-11-26T16:02:45.688205Z  INFO node::user:     Block Time: "Wed, 26 Nov 2025 18:49:03 +0000"

2025-11-26T16:02:54.530132Z  INFO sc_basic_authorship::basic_authorship: 🙌 Starting consensus session on top of parent 0x0dfaf1d789b56b179e4ff0ad0d4e0585806c1b31236606bc939c5deb61c67e54 (#15923)
2025-11-26T16:02:54.532103Z  INFO parachain-system: unincluded_segment_len: 1
2025-11-26T16:02:54.532154Z  INFO parachain-system: included_head: 0x0dfaf1d789b56b179e4ff0ad0d4e0585806c1b31236606bc939c5deb61c67e54
2025-11-26T16:02:54.534758Z  INFO parachain-system: ancestor that will be added to unincluded segment: <wasm:stripped>
2025-11-26T16:02:54.536483Z  INFO sc_basic_authorship::basic_authorship: 🎁 Prepared block for proposing at 15924 (6 ms) hash: 0x4bb47a7cd8af180535ab12cc7a4f8a43f54dddba5ac819eb1e61f03679823051; parent_hash: 0x0dfa…7e54; end: NoMoreTransactions; extrinsics_count: 2
```
</details>

iulianbarbu commented on crates/anvil-polkadot/src/substrate_node/service/mod.rs:
```diff
@@ -49,17 +61,209 @@ pub struct Service {
+    let mut genesis_block_number = anvil_config.get_genesis_number();
+    if let Some(ref fork_url) = anvil_config.eth_rpc_url {
+        // TODO ws is for local host, wss for remote (aka prod)
+        let http_url = fork_url.replacen("https://", "ws://", 1).replacen("http://", "ws://", 1);
```
We could address this todo now, right? It is just a matter of how the eth_rpc_url looks like, so in case of local zombienet testing, we can safely use the http/ws schemes.

iulianbarbu commented on crates/anvil-polkadot/src/substrate_node/service/mod.rs:
```diff
@@ -49,17 +61,209 @@ pub struct Service {
+        // TODO ws is for local host, wss for remote (aka prod)
+        let http_url = fork_url.replacen("https://", "ws://", 1).replacen("http://", "ws://", 1);
+        let fork_choice = anvil_config.fork_choice;
+        let storage_map = std::thread::spawn(move || -> eyre::Result<Result<u64, ()>> {
```
Can we easily make the fn where we run this thread async, to avoid the creation of an std thread with a dedicated tokio runtime in it?

iulianbarbu replied:
Also, 
```suggestion
        let fork_block_number = std::thread::spawn(move || -> eyre::Result<Result<u64, ()>> {
```

iulianbarbu commented on crates/anvil-polkadot/tests/it/forking.rs:
```diff
@@ -591,3 +595,554 @@ async fn test_fork_with_contract_deployment() {
+    let baltathar_address = Address::from(ReviveAddress::new(baltathar.address()));
+
+    // Set initial balances for dev accounts (they may not have balance in the forked chain)
+    let initial_balance = U256::from(100_000_000_000_000_000_000u128); // 100 ether
```
nit: E notation makes it easier to read

iulianbarbu reviewed (commented):
LGTM in general! Ideally we can run these tests without the need to spawn a local zn network manually. This way we can see these tests run in the CI, although I imagine they will take a long time due to the fact our CI is slow. We can leave these as an improvement idea for this branch before merging it in `master`:
1. start zn local network based on zn-sdk
2. upgrade foundry-polkadot CI machines to run the tests faster

dimartiro commented on crates/anvil-polkadot/tests/it/forking.rs:
```diff
@@ -14,6 +14,10 @@ use anvil_polkadot::{
 
+/// Westend Asset Hub zombienet local URL for forking tests
+/// This URL should point to a running zombienet instance
+const WESTEND_ASSET_HUB_URL: &str = "http://127.0.0.1:63982";
```
I'd prefer to keep the real URL since these tests are behind a feature flag (forking-tests) and won't run in CI by default.
Another alternative is to use an ENV, wdyt?

alindima commented on crates/anvil-polkadot/tests/it/forking.rs:
```diff
@@ -591,3 +597,479 @@ async fn test_fork_with_contract_deployment() {
+
+// =============================================================================
+// Tests forking from external Westend Asset Hub zombienet
+// These tests require a running zombienet instance at WESTEND_ASSET_HUB_URL
```
so what's the plan for making these tests runnable in CI?

alindima commented on crates/anvil-polkadot/tests/it/forking.rs:
```diff
@@ -591,3 +595,554 @@ async fn test_fork_with_contract_deployment() {
+    .unwrap();
+    assert!(reverted, "Revert should succeed");
+
+    // Wait for state to settle
```
agreed, need more details here as to why it's needed

alindima commented on crates/anvil-polkadot/tests/it/forking.rs:
```diff
@@ -591,3 +597,479 @@ async fn test_fork_with_contract_deployment() {
+        "Second transaction should succeed"
+    );
+
+    // Verify block number increased
```
missed an assert here

alindima reviewed (commented):
Looking good
